desc/objectives/_coils.py

```diff
@@ -0,0 +1,591 @@
+        from desc.coils import CoilSet, MixedCoilSet, _Coil
+
+        self._dim_f = 0
+
+        def get_dim_f(coilset):
+            """Get dim_f."""
+            if isinstance(coilset, list):
+                [get_dim_f(x) for x in coilset]
+            elif isinstance(coilset, MixedCoilSet):
+                [get_dim_f(x) for x in coilset]
+            elif isinstance(coilset, CoilSet):
+                get_dim_f(coilset.coils)
+            elif isinstance(coilset, LinearGrid):
+                self._dim_f += coilset.num_zeta
+
+        def to_list(coilset):
+            """Turn a MixedCoilSet container into a list of what it's containing."""
+            if isinstance(coilset, list):
+                return [to_list(x) for x in coilset]
+            elif isinstance(coilset, MixedCoilSet):
+                return [to_list(x) for x in coilset]
+            elif isinstance(coilset, CoilSet):
+                # use the same grid/transform for CoilSet
+                return to_list(coilset.coils[0])
+            else:
+                return [coilset]
+
+        is_single_coil = lambda x: isinstance(x, _Coil) and not isinstance(x, CoilSet)
+        # gives structure of coils, e.g. MixedCoilSet(coils, coils) would give a
+        # a structure of [[*, *], [*, *]] if n = 2 coils
+        coil_structure = tree_structure(
+            self.things[0],
+            is_leaf=lambda x: is_single_coil(x),
+        )
+        coil_leaves = tree_leaves(self.things[0], is_leaf=lambda x: is_single_coil(x))
+
+        # check type
+        if isinstance(self._grid, numbers.Integral):
+            self._grid = LinearGrid(N=self._grid, endpoint=False)
+        # all of these cases return a container MixedCoilSet that contains
+        # LinearGrids. i.e. MixedCoilSet.coils = list of LinearGrid
+        if self._grid is None:
+            # map default grid to structure of inputted coils
+            self._grid = tree_map(
+                lambda x: LinearGrid(
+                    N=2 * x.N + 5, NFP=getattr(x, "NFP", 1), endpoint=False
+                ),
+                self.things[0],
+                is_leaf=lambda x: is_single_coil(x),
+            )
+        elif isinstance(self._grid, LinearGrid):
+            # map inputted single LinearGrid to structure of inputted coils
+            self._grid = [self._grid] * len(coil_leaves)
+            self._grid = tree_unflatten(coil_structure, self._grid)
+        else:
+            # this case covers an inputted list of grids that matches the size
+            # of the inputted coils. Can be a 1D list or nested list.
+            flattened_grid = tree_flatten(
+                self._grid, is_leaf=lambda x: isinstance(x, LinearGrid)
+            )[0]
+            self._grid = tree_unflatten(coil_structure, flattened_grid)
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        transforms = tree_map(
```

Comment:
can this be moved to after line 180? For the case where we only have 1 unique grid for a coilset we probably don't have to compute all the extra transforms

Reply:
I'm not sure it could be since tree_map requires self._grid and self.things[0] to be of the same structure, right? so if self._grid is just one of the leaves then tree_map wouldn't work I think